codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java

```diff
@@ -111,123 +112,152 @@ public ChannelFuture writeData(ChannelHandlerContext ctx, int streamId, ByteBuf
             int payloadLength = data.readableBytes() + padding + flags.getPaddingPresenceFieldLength();
             verifyPayloadLength(payloadLength);
 
-            ByteBuf out = ctx.alloc().buffer(FRAME_HEADER_LENGTH + payloadLength);
+            writeFrameHeader(ctx, payloadLength, DATA, flags, streamId, promiseAggregator.newPromise());
```

Comment:
@louiscryan - Yes I agree.  Lets focus on the `DefaultHttp2FrameWriter` in this PR (I am optimizing now).  I opened https://github.com/netty/netty/issues/3390 for this issue.